File: app/src/main/java/com/example/projectowen/MainActivity.java
package com.example.projectowen;

import androidx.appcompat.app.AppCompatActivity;
import androidx.work.WorkManager;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.SeekBar;

import net.ffst.adbpotato.Bridge;

import java.io.IOException;

public class MainActivity extends AppCompatActivity {

    private static class InnerThread extends Thread {

        private boolean running;
        private final MainActivity activity;
        private JoystickApp app;

        private static final int buttonIds[] = new int[]{
                R.id.l0c0, R.id.l1c0, R.id.l2c0,
                R.id.l0c1, R.id.l1c1, R.id.l2c1,
                R.id.l0c2, R.id.l1c2, R.id.l2c2,
                R.id.l0c3, R.id.l1c3, R.id.l2c3,
                R.id.l0c4, R.id.l1c4, R.id.l2c4,
                R.id.l0c5, R.id.l1c5, R.id.l2c5,
                R.id.l0c6, R.id.l1c6, R.id.l2c6,
                R.id.l0c7, R.id.l1c7, R.id.l2c7,
                R.id.l0c8, R.id.l1c8, R.id.l2c8,
        };

        public InnerThread(MainActivity activity, JoystickApp app) {
            this.activity = activity;
            this.app = app;
        }

        @Override
        public void run() {
            running = true;
            while (running) {
                activity.runOnUiThread(() -> {
                    for (int col = 0; col < 9; col++) {
                        for (int level = 0; level < 3; level++) {
                            activity.findViewById(buttonIds[col * 3 + level]).setBackgroundColor(app.getGridState(level, col).backgroundColor);
                        }
                    }
                });
                try {
                    sleep(20); // TODO figure out if we need this and for how long it needs to be
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
            }
        }

        public void end() {
            running = false;
        }
    }

    private InnerThread thread;

    protected void onCreate(Bundle savedInstance) {
        super.onCreate(savedInstance);
        JoystickApp app = ((JoystickApp) getApplication());
        thread = new InnerThread(this, app);
        // cone led control
        findViewById(R.id.ledCone).setOnClickListener((btn) -> {
            try {
                app.bridge.publish("LED Cone", true);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        });
        // cube led control
        findViewById(R.id.ledCube).setOnClickListener((btn) -> {
            try {
                app.bridge.publish("LED Cube", true);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        });
        findViewById(R.id.home).setOnClickListener((btn) -> {
            try {
                app.bridge.publish("Home", true);
            } catch (IOException e){
                throw new RuntimeException(e);
            }
        });
        ((SeekBar)findViewById(R.id.rollers)).setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                try {
                    app.bridge.publish("slider", seekBar.getProgress() / (double) seekBar.getMax() * 2.0 - 1.0);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                seekBar.setProgress(seekBar.getMax() / 2);
                try {
                    app.bridge.publish("slider", 0.0);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }
        });
    }

    @Override
    protected void onDestroy() {
        try {
            thread.end();
            thread.join();
        } catch (InterruptedException ignored) {
        }
        super.onDestroy();
    }

    public void gotoAuto(View v) {
        startActivity(new Intent(MainActivity.this, AutoActivity.class));
    }
}

File: app/src/main/java/com/example/projectowen/MainActivity.kt
package com.example.projectowen

import android.app.Activity
import android.content.pm.ActivityInfo
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.media.Image
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.ConditionVariable
import android.widget.Button
import android.widget.ImageButton
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import java.io.BufferedInputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.net.ServerSocket
import java.net.SocketException

class MainActivity : AppCompatActivity() {
    companion object {
        val buttons = intArrayOf(
            R.id.l0c0, R.id.l0c1, R.id.l0c2, R.id.l0c3, R.id.l0c4, R.id.l0c5, R.id.l0c6, R.id.l0c7, R.id.l0c8,
            R.id.l1c0, R.id.l1c1, R.id.l1c2, R.id.l1c3, R.id.l1c4, R.id.l1c5, R.id.l1c6, R.id.l1c7, R.id.l1c8,
            R.id.l2c0, R.id.l2c1, R.id.l2c2, R.id.l2c3, R.id.l2c4, R.id.l2c5, R.id.l2c6, R.id.l2c7, R.id.l2c8,
        )
    }

    class SocketWorker(port: Int, private var activity: Activity) : Thread() {

        private var socket = ServerSocket(port)
        private var cond = ConditionVariable()
        private var active = 0
        private var intake: Double = 0.0
        private var home = false;
        private var led_cone = false;
        private var led_cube = false;

        private val buttons = intArrayOf(
            R.id.l0c0, R.id.l0c1, R.id.l0c2, R.id.l0c3, R.id.l0c4, R.id.l0c5, R.id.l0c6, R.id.l0c7, R.id.l0c8,
            R.id.l1c0, R.id.l1c1, R.id.l1c2, R.id.l1c3, R.id.l1c4, R.id.l1c5, R.id.l1c6, R.id.l1c7, R.id.l1c8,
            R.id.l2c0, R.id.l2c1, R.id.l2c2, R.id.l2c3, R.id.l2c4, R.id.l2c5, R.id.l2c6, R.id.l2c7, R.id.l2c8,
        )

        var buttonToGroup = HashMap<Int, Pair<Int, Int>>()

        init {
            for(level in 0..2) {
                for (col in 0..8) {
                    buttonToGroup[buttons[level * 9 + col]] = Pair(level, col)
                }
            }
        }

        override fun run() {
            while(true) {
                try {
                    activity.runOnUiThread {
                        val textView = activity.findViewById<TextView>(R.id.connection_status)
                        textView.setText(R.string.not_connected)
                    }

                    println("Waiting for client")
                    val client = socket.accept()
                    println("Client found!")

                    activity.runOnUiThread {
                        val textView = activity.findViewById<TextView>(R.id.connection_status)
                        textView.setText(R.string.connected)
                    }

                    val input = DataInputStream(BufferedInputStream(client.getInputStream()))
                    val output = DataOutputStream(client.getOutputStream())

                    while (true) {
                        cond.block();
                        // buttons
                        if (buttonToGroup.containsKey(active)) {
                            output.writeByte(2)
                            val lc = buttonToGroup[active]
                            if (lc != null) {
                                output.writeByte(lc.first)
                                output.writeByte(lc.second)
                            } else {
                                output.writeByte(1)
                            }
                        } else {
                            output.writeByte(1)
                        }
                        // intake
                        output.writeByte(3)
                        output.writeDouble(intake)

                        // home
                        output.writeByte(4);
                        output.writeBoolean(home);

                        home = false;

                        // led cone
                        output.writeByte(5)
                        output.writeBoolean(led_cone)

                        led_cone = false;

                        // led cube
                        output.writeByte(6)
                        output.writeBoolean(led_cube)

                        led_cube = false;

                        output.flush()
                        cond.close();
                    }
                } catch (exc: SocketException) {
                    // Do nothing
                }
            }
        }

        fun setActive(x: Int) {
            active = x
            cond.open()
        }

        fun setIntake(x: Double) {
            intake = x
            cond.open()
        }

        fun setHome() {
            home = true;
            cond.open();
        }

        fun setCone() {
            led_cone = true;
            cond.open();
        }

        fun setCube() {
            led_cube = true;
            cond.open()
        }
    }

    private fun reset() {
        buttons.forEach {
            val button = findViewById<ImageButton>(it)
            if((button.background as ColorDrawable).color != Color.DKGRAY) {
                button.setBackgroundColor(Color.CYAN)
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE

        val worker = SocketWorker(1235, this)
        worker.start()

        val rollers = findViewById<SeekBar>(R.id.rollers);
        rollers.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(p0: SeekBar?, p1: Int, p2: Boolean) {
                worker.setIntake(((p0?.progress?.toDouble()!!) / (p0.max.toDouble())) * 2.0 - 1.0)
            }

            override fun onStartTrackingTouch(p0: SeekBar?) {

            }

            override fun onStopTrackingTouch(p0: SeekBar?) {
                p0?.setProgress(p0.max / 2, false)
                worker.setIntake(0.0)
            }

        });

        val home_button = findViewById<ImageButton>(R.id.home);

        home_button.setOnClickListener { btn ->
            worker.setHome()
        }

        val cone_button = findViewById<ImageButton>(R.id.ledCone);

        cone_button.setOnClickListener { btn ->
            worker.setCone()
        }

        val cube_button = findViewById<ImageButton>(R.id.ledCube);

        cube_button.setOnClickListener { btn ->
            worker.setCube()
        }

        buttons.forEach {
            val button = findViewById<ImageButton>(it)
            button.setBackgroundColor(Color.DKGRAY)
            button.setOnClickListener { btn ->
                reset()
                if ((btn.background as ColorDrawable).color == Color.CYAN) {
                    btn.setBackgroundColor(Color.DKGRAY)
                } else {
                    btn.setBackgroundColor(Color.RED)
                    worker.setActive(it)
                }
            }
        }
    }
}